lib/serve/functions.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+var _ = require('lodash');
+var chalk = require('chalk');
+var path = require('path');
+var RSVP = require('rsvp');
+var getProjectId = require('../getProjectId');
+var logger = require('../logger');
+var utils = require('../utils');
+var parseTriggers = require('../parseTriggers');
+var EmulatorController;
+var controller;
+
+function _pollOperation(service, operation) {
+  return new RSVP.Promise(function(resolve, reject) {
+    var poll = function() {
+      service.operations.get({
+        name: operation[0].name
+      }, function(err, _operation) {
+        if (err) {
+          reject(err);
+        } else if (_operation.done) {
+        // TODO: handle deploy failures
+          resolve(_operation.response.value);
+        } else {
+          setTimeout(poll, 500);
+        }
+      });
+    };
+    poll();
+  });
+}
+
+function _startEmulator(options) {
+  try {
+    EmulatorController = require('@google-cloud/functions-emulator/src/cli/controller');
+  } catch (err) {
+    var msg = err;
+    if (process.version !== 'v6.9.1') {
+      msg = 'Please use Node version v6.9.1, you have ' + process.version + '\n';
+    }
+    utils.logWarning('Cannot start functions emulator. ' + msg);
+    return RSVP.reject();
+  }
+  if (process.version !== 'v6.9.1') {
+    utils.logWarning('The functions emulator works best with Node version v6.9.1, you have ' + process.version + '\n');
+  }
+  var functionsDir = path.join(options.config.projectDir, options.config.get('functions.source'));
+  controller = new EmulatorController({verbose: true});
+  // TODO: pipe logs to console
+  logger.info(chalk.bold('Functions Directory:'), options.config.get('functions.source'), '\n');
+
+  return controller.start().then(function() {
+    return controller.clear();
+  }).then(function() {
+    return parseTriggers(getProjectId(options), options.instance, functionsDir);
+  // TODO: display helpful message when there are errors from trigger parsing
+  }).then(function(triggers) {
+    var promises = _.map(triggers, function(trigger) {
+      if (trigger.httpsTrigger) {
+        return controller.deploy(trigger.name, {
```

Comment:
Hoo boy. The fact that you have to asynchronously "deploy" functions to the emulator seems very overcomplicated to me. Not your design, I know, but it's making the output very messy. This is what happened when I just tried to run it:

```
Project Directory: /usr/local/google/home/bleigh/code/firesale
Starting Cloud Functions Emulator...
Functions Directory: functions
Starting Firebase development hosting server...

⚠  Port 5000 is not available. Trying another port...
Starting Firebase development hosting server...

Public Directory: public

Hosting server listening at: http://localhost:5001
Error: Timeout waiting for emulator start
    at client.testConnection.catch (/usr/local/google/home/bleigh/code/firebase-tools/node_modules/@google-cloud/functions-emulator/src/cli/controller.js:230:17)
    at process._tickCallback (internal/process/next_tick.js:103:7)
```

I think some more attention needs to be paid to making the console output clean on startup -- for instance it was very unclear to me that I was waiting on the functions emulator to boot (and then it timed out which I don't know why it did).